Implement Subscriptions API Filters in MTChannelBroker

devguyio opened the pull request:
Part of #5204 

## TODO
- [x] Implement Prefix filter
- [x] Implement Suffix filter
- [x] Implement CESQL filter
- [x] More CESQL unit tests
- [x] Update mtbroker filter handler with parsing `Filters` field
- [x] e2e tests 
- [x] docs

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

<!-- Please categorize your changes:
- :gift: Add new feature
- :bug: Fix bug
- :broom: Update or clean up current behavior
- :wastebasket: Remove feature or internal logic
-->

- Implement SubscriptionsAPI Filters

### Pre-review Checklist

<!-- If these boxes are not checked, you will be asked to complete these requirements or explain why they do not apply to your PR. -->

- [x] **At least 80% unit test coverage**
- [x] **E2E tests** for any new behavior
- [x] **Docs PR** for any user-facing impact
- [ ] ~~**Spec PR** for any new API feature~~
- [ ] ~~**Conformance test** for any change to the spec~~

**Release Note**

<!--
:page_facing_up: If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release note is needed.
-->

```release-note
- 🎁 Add new `new-trigger-filters` experimental feature. When enabled, Triggers support a new `filters` field that conforms to the filters API field defined in the [`CloudEvents Subscriptions API`](https://github.com/cloudevents/spec/blob/main/subscriptions/spec.md#324-filters). It allows you to specify a set of powerful filter expressions, where each expression evaluates to either true or false for each event.
```


**Docs**

<!--
:book: If this change has user-visible impact, link to an issue or PR in
https://github.com/knative/docs.
-->
https://github.com/knative/docs/pull/4635

pierDipi commented:
/cc

pierDipi reviewed (commented):
Why another package `subscriptionsapi`?

types like `eventfilter.Filters` were added for the exact same feature.

devguyio commented:
> Why another package `subscriptionsapi`?
> 
> types like `eventfilter.Filters` were added for the exact same feature.

Can you clarify more what's the other approach you're suggesting? This is a nested package inside `eventfilter` . So we end up with:
- `eventfilter.attribute` nested package which implements an exact match across a set of attributes (existing `Trigger.Spec.Filter`).
- `eventfilter.subscriptionsapi` added in this PR, which is an implementation of the SubscriptionsAPI filter dialect that's using the same `Filter` interface and reusing some of the existing types like `Filters` and the lookup of attributes in `eventfilter.attributes` package.

matzew commented:
> Why another package `subscriptionsapi`?

I like the separation that Ahmed lists out here for the new feature/API, and relationship to the filter

pierDipi commented:
I was referring to the fact the type `eventfilter.Filters` is exactly the same as the new `all` filter that is in the new `subscriptionsapi` package and the attribute filter is exactly the same as the existing attribute filter.

Example types that aren't doing anything useful:

`all_filter`:
```golang
type allFilter eventfilter.Filters

// NewAllFilter returns an event filter which passes if all the contained filters pass
func NewAllFilter(filters ...eventfilter.Filter) eventfilter.Filter {
	return append(allFilter{}, filters...)
}

func (f allFilter) Filter(ctx context.Context, event cloudevents.Event) eventfilter.FilterResult {
	return eventfilter.Filters(f).Filter(ctx, event)
}

var _ eventfilter.Filter = allFilter{}

```

`exact_filter`:

```golang
type exactFilter struct {
	attribute, value string
	attrsFilter      eventfilter.Filter
}

// NewExactFilter returns an event filter which passes if value exactly matches the value of the context
// attribute in the CloudEvent.
func NewExactFilter(attribute, value string) eventfilter.Filter {
	return &exactFilter{
		attribute: attribute,
		value:     value,
		// we're creating this filter to leverage the same filter logic of the existing attributes filter
		attrsFilter: attributes.NewAttributesFilter(map[string]string{attribute: value}),
	}
}

func (f *exactFilter) Filter(ctx context.Context, event cloudevents.Event) eventfilter.FilterResult {
	logger := logging.FromContext(ctx)
	logger.Debugw("Performing an exact match ", zap.String("attribute", f.attribute), zap.String("value", f.value),
		zap.Any("event", event))
	return f.attrsFilter.Filter(ctx, event)
}
```

Do we really need these 2 new types?

Why not a more flattened package for filters that helps with discoverability?
I just need to type `eventfilter.` in my IDE to get the list of filters

devguyio commented:
> Do we really need these 2 new types?
> 
> Why not a more flattened package for filters that helps with discoverability? I just need to type `eventfilter.` in my IDE to get the list of filters

I see your points and I had to do some thinking there when I added them exactly for that reason. Specially with `allFilter` and `Filters`. Here's my thought process.

1. The existing structure was `eventfilter` that has the more generic types: the interface, the results and some helpers.
2. The `Filters` type was used historically for the v1alpha1 Trigger APIs that had type and source as filter fields separate from extensions. When that was removed, the `Filters` type wasn't and was being used as a slice with a single filter.
3. My goal is to separate the old filters from the new ones because it's still an experimental feature and we'll end up with either of them. 

So I removed the `Filters` type which is no longer needed as mentioned above.

odacremolbap commented on pkg/eventfilter/subscriptionsapi/exact_filter.go:
```diff
@@ -0,0 +1,55 @@
+	"knative.dev/eventing/pkg/eventfilter"
+)
+
+type exactFilter struct {
+	attribute, value string
+	attrsFilter      eventfilter.Filter
+}
+
+// NewExactFilter returns an event filter which passes if value exactly matches the value of the context
+// attribute in the CloudEvent.
+func NewExactFilter(attribute, value string) eventfilter.Filter {
+	return &exactFilter{
+		attribute: attribute,
+		value:     value,
+		// we're creating this filter to leverage the same filter logic of the existing attributes filter
+		attrsFilter: attributes.NewAttributesFilter(map[string]string{attribute: value}),
+	}
+}
+
+func (filter *exactFilter) Filter(ctx context.Context, event cloudevents.Event) eventfilter.FilterResult {
+	if filter == nil || filter.attribute == "" || filter.value == "" {
+		return eventfilter.NoFilter
+	}
+	logger := logging.FromContext(ctx)
+	logger.Debugw("Performing an exact match ", zap.String("attribute", filter.attribute), zap.String("value", filter.value),
+		zap.Any("event", event))
+	return filter.attrsFilter.Filter(ctx, event)
+}
```
Feels like providing a sub-case for the Attributes filter, I wouldn't even create a type for it.

```suggestion
// NewExactFilter returns an event filter which passes if value exactly matches the value of the context
// attribute in the CloudEvent.
func NewExactFilter(attribute, value string) eventfilter.Filter {
	if attribute == "" || value == "" {
		return nil
	}

	// we're creating this filter to leverage the same filter logic of the existing attributes filter
	return attributes.NewAttributesFilter(map[string]string{attribute: value})
}
```

devguyio replied:
I created a type for two reasons:
1. I want to provide the same debugging logs as the other filters. Given such complicated feature, I believe debug logs for filters might be useful.
2. I wanted to be explicit about the difference between the two filters, this one accepts only one, and it's visible in the factory function signature as well as the type itself.

odacremolbap commented on pkg/eventfilter/subscriptionsapi/prefix_filter.go:
```diff
@@ -0,0 +1,63 @@
+// NewPrefixFilter returns an event filter which passes if the value of the context
+// attribute in the CloudEvent is prefixed with prefix.
+func NewPrefixFilter(attribute, prefix string) eventfilter.Filter {
+	return &prefixFilter{
+		attribute: attribute,
+		prefix:    prefix,
+	}
```
nit, saving some memory.
```suggestion
	if attribute == "" || prefix == "" {
		return nil
	}
	return &prefixFilter{
		attribute: attribute,
		prefix:    prefix,
	}
```

odacremolbap commented on pkg/eventfilter/subscriptionsapi/suffix_filter.go:
```diff
@@ -0,0 +1,63 @@
+// NewSuffixFilter returns an event filter which passes if the value of the context
+// attribute in the CloudEvent ends with suffix.
+func NewSuffixFilter(attribute, suffix string) eventfilter.Filter {
+	return &suffixFilter{
+		attribute: attribute,
+		suffix:    suffix,
+	}
```
```suggestion
	if attribute == "" || suffix == "" {
		return nil
	}

	return &suffixFilter{
		attribute: attribute,
		suffix:    suffix,
	}
```

odacremolbap commented on pkg/eventfilter/subscriptionsapi/exact_filter.go:
```diff
@@ -0,0 +1,55 @@
+
+// NewExactFilter returns an event filter which passes if value exactly matches the value of the context
+// attribute in the CloudEvent.
+func NewExactFilter(attribute, value string) eventfilter.Filter {
```
From the spec for `exact`, `prefix` and `suffix`:
```
The attribute name and value specified in the filter express MUST NOT be empty strings.
```

Shouldn't the constructor return an error as the second parameter just as the `sql` does?

odacremolbap reviewed (commented):
Although I added a bunch of comments they are all very small and opinionated (sorry for the noise). Overall looks very good.

devguyio commented:
I added e2e tests. Here's an output snippet 
```
--- PASS: TestMTChannelBrokerNewTriggerFilters (66.58s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/Prerequisite (3.05s)
        --- PASS: TestMTChannelBrokerNewTriggerFilters/Prerequisite/Multi-tenant_channel-based_broker/Setup/Install_broker_"default" (0.02s)
        --- PASS: TestMTChannelBrokerNewTriggerFilters/Prerequisite/Multi-tenant_channel-based_broker/Requirement/Broker_is_ready (3.02s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Install_trigger_subscriber (3.25s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Install_trigger (0.04s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Wait_for_trigger_to_become_ready (3.03s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Install_matched_event_sender (3.06s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Install_unmatched_event_sender (3.06s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Exact_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (0.00s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Exact_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Install_trigger_subscriber (3.06s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Install_trigger (0.02s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Wait_for_trigger_to_become_ready (3.02s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Install_matched_event_sender (3.11s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Install_unmatched_event_sender (3.05s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Prefix_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (0.00s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Prefix_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Install_trigger_subscriber (3.05s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Install_trigger (0.01s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Wait_for_trigger_to_become_ready (3.01s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Install_matched_event_sender (3.05s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Install_unmatched_event_sender (3.05s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Suffix_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (0.00s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/Suffix_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Install_trigger_subscriber (3.05s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Install_trigger (0.02s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Wait_for_trigger_to_become_ready (3.01s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Install_matched_event_sender (3.10s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Install_unmatched_event_sender (3.09s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (0.00s)
    --- PASS: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
PASS
ok      knative.dev/eventing/test/experimental  66.663s
```

devguyio commented:
@pierDipi @odacremolbap review comments addressed and PR is ready for review

devguyio commented:
There's a failing e2e test
```
    --- FAIL: TestMTChannelBrokerNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
```

devguyio commented:
Fixed, the test predicate for the `prefix` was wrong.

pierDipi commented on test/experimental/features/new_trigger_filters/filters.go:
```diff
@@ -0,0 +1,123 @@
+		}
+
+		f.Setup("Install trigger", trigger.Install(triggerName, brokerName, cfg...))
+		f.Setup("Wait for trigger to become ready", trigger.IsReady(triggerName))
+
+		f.Setup("Install matched event sender", func(ctx context.Context, t feature.T) {
+			u, err := broker.Address(ctx, brokerName)
+			if err != nil || u == nil {
+				t.Error("failed to get the address of the broker", brokerName, err)
+			}
+			eventshub.Install(matchedSender, eventshub.StartSenderURL(u.String()), eventshub.InputEvent(matchedEvent))(ctx, t)
+		})
+
+		f.Setup("Install unmatched event sender", func(ctx context.Context, t feature.T) {
+			u, err := broker.Address(ctx, brokerName)
+			if err != nil || u == nil {
+				t.Error("failed to get the address of the broker", brokerName, err)
+			}
+			eventshub.Install(unmatchedSender, eventshub.StartSenderURL(u.String()), eventshub.InputEvent(unmatchedEvent))(ctx, t)
+		})
```
Since the spec specify as must that only events that are sent after a trigger is ready must be sent to the subscriber, this might be racy if the sender starts before the trigger is ready, so I'd move `install ... event sender` to the requirement phase.
```suggestion
		f.Setup("Wait for trigger to become ready", trigger.IsReady(triggerName))

		f.Requirement("Install matched event sender", func(ctx context.Context, t feature.T) {
			u, err := broker.Address(ctx, brokerName)
			if err != nil || u == nil {
				t.Error("failed to get the address of the broker", brokerName, err)
			}
			eventshub.Install(matchedSender, eventshub.StartSenderURL(u.String()), eventshub.InputEvent(matchedEvent))(ctx, t)
		})

		f.Requirement("Install unmatched event sender", func(ctx context.Context, t feature.T) {
			u, err := broker.Address(ctx, brokerName)
			if err != nil || u == nil {
				t.Error("failed to get the address of the broker", brokerName, err)
			}
			eventshub.Install(unmatchedSender, eventshub.StartSenderURL(u.String()), eventshub.InputEvent(unmatchedEvent))(ctx, t)
		})
```

devguyio replied:
good catch 👍🏽

pierDipi commented on test/experimental/features/new_trigger_filters/filters.go:
```diff
@@ -0,0 +1,123 @@
+	unmatchedEvent.SetType("com.wrong.type")
+	unmatchedEvent.SetSource("com.wrong.source")
+
+	var features []feature.Feature
```
```suggestion
	features := make([]feature.Feature, 0, 16)
```

pierDipi commented on pkg/eventfilter/subscriptionsapi/cesql_filter.go:
```diff
@@ -0,0 +1,73 @@
+	if filter == nil || filter.rawExpression == "" {
+		return eventfilter.NoFilter
+	}
+	logger := logging.FromContext(ctx)
+	logger.Debugw("Performing a CESQL match ", zap.String("expression", filter.rawExpression), zap.Any("event", event))
+
+	res, err := filter.parsedExpression.Evaluate(event)
+
+	if err != nil {
+		logger.Debugw("Error evaluating expression on event.", zap.String("expression", filter.rawExpression), zap.Error(err))
+		return eventfilter.FailFilter
+	}
+
+	if !res.(bool) {
+		logger.Debugw("CESOL match failed.", zap.String("expression", filter.rawExpression), zap.Any("event", event))
+		return eventfilter.FailFilter
+	}
```
I like the use of the logger in the filter functions and I have a question.
Is this logger configured to have fields linking back to the trigger originating a given filter type?

devguyio replied:
Fixed.

pierDipi reviewed (commented):
LGTM, some minor comments, and one around rekt that I believe should be fixed before merging.

pierDipi commented on pkg/broker/filter/filter_handler.go:
```diff
@@ -333,15 +334,84 @@ func (h *Handler) getTrigger(ref path.NamespacedNameUID) (*eventingv1.Trigger, e
+}
+
+func materializeFiltersList(ctx context.Context, filters []eventingv1.SubscriptionsAPIFilter) []eventfilter.Filter {
+	materializedFilters := make([]eventfilter.Filter, len(filters))
```
This needs 0 as starting index:
```suggestion
	materializedFilters := make([]eventfilter.Filter, 0, len(filters))
```

odacremolbap commented on pkg/eventfilter/subscriptionsapi/any_filter.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+func (filter anyFilter) Filter(ctx context.Context, event cloudevents.Event) eventfilter.FilterResult {
+	if filter == nil {
+		return eventfilter.NoFilter
+	}
```
mini nit: superfluous, can be removed
```suggestion
```

odacremolbap reviewed (commented):
I am not sure if I might be missing something regarding enabling experimental features + filtering. I looked for `NewTriggerFilters` at the whole code and the only `if` I found was at the filter validation while I was expecting it also at the filter handler.

(See comment + suggestion at the review)

devguyio commented on pkg/broker/filter/filter_handler.go:
```diff
@@ -333,15 +334,84 @@ func (h *Handler) getTrigger(ref path.NamespacedNameUID) (*eventingv1.Trigger, e
+	if filterSpec.Filter != nil {
+		return applyAttributesFilter(ctx, filterSpec.Filter, event)
+	} else if len(filterSpec.Filters) > 0 {
+		return applySubscriptionsAPIFilters(ctx, filterSpec.Filters, event)
 	}
-	var filters eventfilter.Filters
-	if filter.Attributes != nil && len(filter.Attributes) != 0 {
-		filters = append(filters, attributes.NewAttributesFilter(filter.Attributes))
+	return eventfilter.NoFilter
```
That's a great catch, I don't know how I missed that 🙄 . Thanks a lot @odacremolbap , yes you're 100% correct. 🙏🏽 for the thorough review 👏🏽

devguyio replied:
Docs are correct, implementation was wrong and fixed. Context: https://github.com/knative/eventing/pull/5995#issuecomment-1015414489

devguyio commented:
@odacremolbap To clarify for the case of both old and new filters existing, the feature track and the discussions concluded on the new one overriding the old. Reasoning was is that this won't force the users to "delete" the old filter to try out the effect of the new filters. They can add the new filters to an existing trigger, play around with it, and when the want they can just remove it and the old filters will kick it again.

As for the feature flag name, that was also discussed at length and the reason is to emphasize for users that there is "a new filters syntax" (without requiring any context or awareness of the subscriptions API ) , and this will live only through the life of the experimental feature, after graduation they will just be "filters" .

devguyio commented on pkg/broker/filter/filter_handler.go:
```diff
@@ -333,15 +334,84 @@ func (h *Handler) getTrigger(ref path.NamespacedNameUID) (*eventingv1.Trigger, e
+	if len(filter.Prefix) > 0 {
+		// The webhook validates that this map has only a single key:value pair.
+		for attribute, prefix := range filter.Exact {
+			materializedFilter, err = subscriptionsapi.NewPrefixFilter(attribute, prefix)
+			if err != nil {
+				logging.FromContext(ctx).Debugw("Invalid prefix expression", zap.String("attribute", attribute), zap.String("prefix", prefix), zap.Error(err))
+				return nil
+			}
+		}
+	}
+	if len(filter.Suffix) > 0 {
+		// The webhook validates that this map has only a single key:value pair.
+		for attribute, suffix := range filter.Exact {
+			materializedFilter, err = subscriptionsapi.NewSuffixFilter(attribute, suffix)
+			if err != nil {
+				logging.FromContext(ctx).Debugw("Invalid suffix expression", zap.String("attribute", attribute), zap.String("suffix", suffix), zap.Error(err))
+				return nil
+			}
+		}
+	}
+	if len(filter.All) > 0 {
+		materializedFilter = subscriptionsapi.NewAllFilter(materializeFiltersList(ctx, filter.All)...)
 	}
-	return filters.Filter(ctx, event)
+	if len(filter.Any) > 0 {
+		materializedFilter = subscriptionsapi.NewAnyFilter(materializeFiltersList(ctx, filter.All)...)
+
+	}
+	if filter.Not != nil {
+		materializedFilter = subscriptionsapi.NewNotFilter(materializeSubscriptionsAPIFilter(ctx, *filter.Not))
+	}
+	if filter.SQL != "" {
+		f, err := subscriptionsapi.NewCESQLFilter(filter.SQL)
+		if err != nil {
+			// This is weird, CESQL expression should be validated when Trigger's are created.
+			logging.FromContext(ctx).Debugw("Found an Invalid CE SQL expression", zap.String("expression", filter.SQL))
+			return nil
+		}
+		materializedFilter = f
+	}
```
ack, that if logic was there when I didn't come to the realization that only one should be provided as per spec. 👍🏽

devguyio commented on test/experimental/config/features.yaml:
```diff
@@ -26,3 +26,4 @@ data:
   delivery-retryafter: "enabled"
   delivery-timeout: "enabled"
   strict-subscriber: "disabled"
+  new-trigger-filters: "enabled"
```
The name will last only while it's an experimental feature, and will be retired after graduation. It'll be just "filters".  Check the comment below.

devguyio commented:
@odacremolbap @pierDipi comments addressed.

odacremolbap commented:
sweet!, 
holding for @pierDipi definitive comments.

/hold

devguyio commented:
/retest

devguyio commented:
/retest-required

pierDipi commented:
Woot! Thanks for this feature!

/unhold

devguyio commented:
/retest-required

devguyio commented:
/retest-required
